pennylane/ops/op_math/prod.py

```diff
@@ -0,0 +1,210 @@
+
+    @property
+    def num_wires(self):
+        return len(self.wires)
+
+    @property
+    def is_hermitian(self, run_check=True):  # TODO: cache this value, this check is expensive!
+        """check if the product operator is hermitian"""
+        if run_check:
+            mat = self.matrix()
+            adj_mat = math.conj(math.transpose(mat))
+            if math.allclose(mat, adj_mat):
+                return True
+            return False
+        raise qml.operation.IsHermitianUndefinedErrors
+
+    def decomposition(self):
+        """decomposition of the operator into a product of operators."""
+        if qml.queuing.QueuingContext.recording():
+            return [qml.apply(op) for op in self.factors]
+        return list(self.factors)
+
+    @property
+    def eigendecomposition(self):
+        r"""Return the eigendecomposition of the matrix specified by the operator.
+
+        This method uses pre-stored eigenvalues for standard observables where
+        possible and stores the corresponding eigenvectors from the eigendecomposition.
+
+        It transforms the input operator according to the wires specified.
+
+        Returns:
+            dict[str, array]: dictionary containing the eigenvalues and the eigenvectors of the operator
+        """
+        Hmat = self.matrix()
+        Hmat = math.to_numpy(Hmat)
+        Hkey = tuple(Hmat.flatten().tolist())
+        if Hkey not in self._eigs:
+            w, U = np.linalg.eigh(Hmat)
+            self._eigs[Hkey] = {"eigvec": U, "eigval": w}
+
+        return self._eigs[Hkey]
+
+    def diagonalizing_gates(self):
+        r"""Sequence of gates that diagonalize the operator in the computational basis.
+
+        Given the eigendecomposition :math:`O = U \Sigma U^{\dagger}` where
+        :math:`\Sigma` is a diagonal matrix containing the eigenvalues,
+        the sequence of diagonalizing gates implements the unitary :math:`U`.
+
+        The diagonalizing gates rotate the state into the eigenbasis
+        of the operator.
+
+        A ``DiagGatesUndefinedError`` is raised if no representation by decomposition is defined.
+
+        .. seealso:: :meth:`~.Operator.compute_diagonalizing_gates`.
+
+        Returns:
+            list[.Operator] or None: a list of operators
+        """
+
+        eigen_vectors = self.eigendecomposition["eigvec"]
+        return [qml.QubitUnitary(eigen_vectors.conj().T, wires=self.wires)]
+
+    def eigvals(self):
+        r"""Return the eigenvalues of the specified operator.
+
+        This method uses pre-stored eigenvalues for standard observables where
+        possible and stores the corresponding eigenvectors from the eigendecomposition.
+
+        Returns:
+            array: array containing the eigenvalues of the operator
+        """
+        return self.eigendecomposition["eigval"]
+
+    def matrix(self, wire_order=None):
+        """Representation of the operator as a matrix in the computational basis."""
```

Comment:
Is this very time consuming? Maybe we can cache it in a _matrix attribute. And do something like:
```python
if self._matrix is not None:
    return self._matrix
```
And then set `self._matrix` to None every time one of the factors is updated.

Reply:
The one problem with this is that operators are currently mutable.

I'd like to make `Operator`'s immutable, but that will take some work.

Reply:
We can't "simply" cache the matrix because it depends on the `wire_order` parameter. For this reason the same operator would have multiple matrix representations depending on if we add more wires or reorder them.